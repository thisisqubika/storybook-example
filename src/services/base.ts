export type BaseURL = URL | string;

export type FetchResource = URL | string | string[] | {
  pathname: string | string[];
  searchParams?: Record<string, unknown>;
}; // type FetchResource

export const MIME_TYPES = {
  json: 'application/json',
}; // const MIME_TYPES

export interface FetchOptions extends Omit<RequestInit, 'body'> {
  body?: unknown;
} // interface FetchOptions

export class FetchError extends Error {
  request: Request;
  response?: Response;

  constructor(message: string, request: Request, response?: Response) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.request = request;
    this.response = response;
  }
} // class FetchError

export interface ApiArgs {
  baseURL: BaseURL;
  defaultHeaders?: HeadersInit;
} // interface ApiArgs

export const DEFAULT_HEADERS: HeadersInit = {
  Accept: MIME_TYPES.json,
  'Content-Type': MIME_TYPES.json,
}; // const DEFAULT_HEADERS

/** Base class for the representation of APIs used by the frontend.
*/
export class Api {
  _baseURL: BaseURL;
  _defaultHeaders?: HeadersInit;

  constructor({
    baseURL,
    defaultHeaders,
  }: ApiArgs) {
    this._baseURL = baseURL;
    this._defaultHeaders = defaultHeaders ?? DEFAULT_HEADERS;
  } // constructor

  /** Creates a URL from a given resource. Uses this object's `baseURL`.
  */
  url(resource: FetchResource) {
    if (resource instanceof URL) {
      return resource;
    }
    if (Array.isArray(resource) || typeof resource === 'string') {
      resource = { pathname: resource };
    }
    const { pathname, searchParams = {} } = resource;
    resource = new URL(
      typeof pathname === 'string' ? pathname
        : `/${pathname.map(encodeURIComponent).join('/')}`,
      this._baseURL,
    );
    for (const [paramName, paramValue] of Object.entries(searchParams)) {
      if (Array.isArray(paramValue)) {
        paramValue.forEach(
          (v) => resource.searchParams.append(paramName, v),
        );
      } else if (paramValue !== undefined) {
        resource.searchParams.set(paramName, `${paramValue}`);
      }
    }
    return resource;
  } // url

  /** Creates a Request instance for the fetch method.
  */
  request(resource: FetchResource, options?: FetchOptions) {
    const url = this.url(resource);
    options = { ...options };
    if (options.body !== undefined) {
      options.body = JSON.stringify(options.body);
    }
    if (!options.headers) {
      options.headers = { ...this._defaultHeaders };
    }
    return new Request(url, options as RequestInit);
  } // request

  /** Wrapper for the standard `fetch` function. It may take any data as `body`,
   * which will be turned to JSON. It also fails when the response status is not
   * 2xx. Resolves to the parsed response.
  */
  async fetch<D>(resource: FetchResource, options?: FetchOptions): Promise<D> {
    const request = this.request(resource, options as RequestInit);
    let message, response;
    try {
      response = await globalThis.fetch(request);
      if (response.ok) {
        const payload = response.status === 204 ? null : await response.json();
        return payload;
      }
      message = `Request failed with ${response.status}: ${
        JSON.stringify(response.statusText)}!`;
    } catch (error) {
      message = (error instanceof Error) ? error.message : `${error}`;
    }
    throw new FetchError(message, request, response);
  } // fetch

  /** Synonym for the `fetch` method using the `DELETE` HTTP method.
  */
  async delete<D>(resource: FetchResource, options?: FetchOptions): Promise<D> {
    return this.fetch(resource, { ...options, method: 'DELETE' });
  } // delete

  /** Synonym for the `fetch` method using the `GET` HTTP method.
  */
  async get<D>(resource: FetchResource, options?: FetchOptions): Promise<D> {
    return this.fetch(resource, { ...options, method: 'GET' });
  } // get

  /** Synonym for the `fetch` method using the `PATCH` HTTP method.
  */
  async patch<D>(resource: FetchResource, options?: FetchOptions): Promise<D> {
    return this.fetch(resource, { ...options, method: 'PATCH' });
  } // patch

  /** Synonym for the `fetch` method using the `POST` HTTP method.
  */
  async post<D>(resource: FetchResource, options?: FetchOptions): Promise<D> {
    return this.fetch(resource, { ...options, method: 'POST' });
  } // post

  /** Synonym for the `fetch` method using the `PUT` HTTP method.
  */
  async put<D>(resource: FetchResource, options?: FetchOptions): Promise<D> {
    return this.fetch(resource, { ...options, method: 'PUT' });
  } // put

} // class Api
