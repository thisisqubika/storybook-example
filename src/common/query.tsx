import {
  QueryClient, QueryClientConfig, QueryClientProvider,
} from '@tanstack/react-query';

export function queryClientConfig(
  config?: QueryClientConfig,
): QueryClientConfig {
  return {
    ...config,
    defaultOptions: {
      ...config?.defaultOptions,
      queries: {
        refetchOnWindowFocus: false,
        retry: 1,
        ...config?.defaultOptions?.queries,
      },
    },
  };
} // function newQueryClient

export interface QueryClientWrapperProps {
  config?: QueryClientConfig;
  children: JSX.Element;
} // interface ThemeWrapperProps

export function QueryClientWrapper({
  config,
  children,
}: QueryClientWrapperProps) {
  const queryClient = new QueryClient(queryClientConfig(config));
  return (
    <QueryClientProvider client={queryClient}>
      {children}
    </QueryClientProvider>
  );
} // function QueryClientWrapper

