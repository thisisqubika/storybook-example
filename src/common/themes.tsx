// Check out <https://zenoo.github.io/mui-theme-creator/>.
import {
  createTheme, CssBaseline, ThemeProvider, ThemeOptions,
} from '@mui/material';

export const themeOptions: ThemeOptions = {
  palette: {
    mode: 'light',
    primary: {
      main: '#812020',
    },
    secondary: {
      main: '#cc8333',
    },
    background: {
      default: '#e0cab8',
      paper: '#efe8e2',
    },
    error: {
      main: '#fb3636',
    },
    info: {
      main: '#025e90',
    },
    warning: {
      main: '#e49e01',
    },
    success: {
      main: '#2b9430',
    },
  },
  typography: {
    fontFamily: 'Lora',
    fontSize: 13,
  },
}; // const themeOptions

export interface ThemeWrapperProps {
  children: JSX.Element;
} // interface ThemeWrapperProps

export function ThemeWrapper({
  children,
}: ThemeWrapperProps) {
  return (
    <ThemeProvider theme={createTheme(themeOptions)}>
      <CssBaseline />
      {children}
    </ThemeProvider>
  );
} // function ThemeWrapper
