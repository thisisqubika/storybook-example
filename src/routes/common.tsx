import { isRouteErrorResponse, useRouteError } from 'react-router-dom';

export const PATHS = {
  books: 'books',
}; // const PATHS

export function ErrorBoundary() {
  const error: unknown = useRouteError();
  return isRouteErrorResponse(error) ? (
    <h1>{error.status} {error.statusText}</h1>
  ) : (
    <h1>{(error as Error)?.message ?? error}</h1>
  );
} // function ErrorBoundary
