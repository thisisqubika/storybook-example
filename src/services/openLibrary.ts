import { useQuery } from "@tanstack/react-query";
import { Api } from "./base";
import { type OpenLibraryBookSearch } from "./openLibrary.types";

export const baseURL = new URL('https://openlibrary.org');

export interface OpenLibraryQueryParams {
  fields?: string;
  lang?: string;
  limit?: number;
  offset: number;
  q: string;
} // interface OpenLibraryQueryParams

export class OpenLibraryApi extends Api {
  constructor() {
    super({
      baseURL,
      defaultHeaders: {},
    });
  } // constructor

  bookCoverImage(id: number | string) {
    return new URL(`https://covers.openlibrary.org/b/id/${
      encodeURIComponent(id)}-M.jpg`);
  } // bookCoverImage

  async searchBooks(params: OpenLibraryQueryParams) {
    return this.get<OpenLibraryBookSearch>({
      pathname: 'search.json', 
      searchParams: {
        fields: 'author_key,author_name,cover_i,key,language,publish_date,publisher,title',
        lang: 'en',
        limit: 10,
        ...params,
      },
    });
  } // searchBooks

} // class OpenLibraryApi

const openLibraryApi = new OpenLibraryApi();

export default openLibraryApi;

export function useFetchBookSearch(params: OpenLibraryQueryParams) {
  return useQuery({
    queryKey: ['search.json', params],
    queryFn: () => openLibraryApi.searchBooks(params),
  });
} // function useFetchBookSearch
