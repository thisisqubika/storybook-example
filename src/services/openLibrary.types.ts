/** Subset of all fields.
*/
export interface OpenLibraryBook {
  author_key?: string[];
  author_name?: string[];
  cover_i?: number;
  key: string;
  language?: string[];
  publish_date?: string[];
  publisher?: string[];
  title: string;
} // interface OpenLibraryBook

export interface OpenLibraryBookSearch {
  docs: OpenLibraryBook[];
  numFound: number;
  num_found: number;
  numFoundExact: boolean;
  offset: number | null;
  q: string;
  start: number;
} // interface OpenLibraryBookSearch