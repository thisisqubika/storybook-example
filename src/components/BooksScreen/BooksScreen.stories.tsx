import type { Meta, StoryObj } from '@storybook/react';
import { http, HttpResponse } from 'msw';
import { ThemeWrapper } from '../../common/themes';
import { QueryClientWrapper } from '../../common/query';
import { fakeSearch } from '../../../test/fixtures/openLibrary';
import { baseURL } from '../../services/openLibrary';
import { BooksScreen } from '.';

export default {
  component: BooksScreen,
  decorators: [
    (Story) => (
      <ThemeWrapper>
        <QueryClientWrapper>
          <Story />
        </QueryClientWrapper>
      </ThemeWrapper>
    ),
  ]
} satisfies Meta; // meta

export const Base: StoryObj = {
  render() {
    return (
      <BooksScreen />
    );
  },
}; // const Base

export const MockedAPI_Success: StoryObj = {
  parameters: {
    msw: {
      handlers: [
        http.get(`${baseURL}search.json`, ({ request }) => {
          const requestURL = new URL(request.url);
          const q = requestURL.searchParams.get('q');
          return HttpResponse.json(fakeSearch(q ?? ''));
        }),
      ],
    },
  }, // parameters

  render() {
    return (
      <BooksScreen />
    );
  },
}; // const MockedAPI_Success

export const MockedAPI_Error: StoryObj = {
  parameters: {
    msw: {
      handlers: [
        http.get(`${baseURL}search.json`, () => {
          return HttpResponse.json({}, { status: 404 });
        }),
      ],
    },
  }, // parameters

  render() {
    return (
      <BooksScreen />
    );
  },
}; // const MockedAPI_Error
