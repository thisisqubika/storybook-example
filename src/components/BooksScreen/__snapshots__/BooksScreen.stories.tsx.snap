// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BooksScreen Base smoke-test 1`] = `
<div class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
     data-testid="PageLayout"
>
  <div class="MuiBox-root css-1o0wkzj">
    <div class="MuiBox-root css-i9gxme">
      <header class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-eoy4xr-MuiPaper-root-MuiAppBar-root"
              data-testid="Header"
      >
        <div class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root">
          <h3 class="MuiTypography-root MuiTypography-h3 css-e6ke5q-MuiTypography-root"
              data-testid="Logo"
          >
            Books
          </h3>
          <div class="css-r7zuen">
            <div class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
                 data-testid="SearchInput"
            >
              <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1h4b0e9-MuiInputBase-root-MuiInput-root">
                <input aria-invalid="false"
                       id=":r0:"
                       placeholder="Search..."
                       type="text"
                       data-testid="SearchInput.input"
                       class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
                       value="React web"
                >
                <div class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root">
                  <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1m1cnyu-MuiButtonBase-root-MuiIconButton-root"
                          tabindex="0"
                          type="button"
                          data-testid="SearchInput.IconButton"
                  >
                    <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-6cvc6r-MuiSvgIcon-root"
                         focusable="false"
                         aria-hidden="true"
                         viewbox="0 0 24 24"
                         data-testid="SearchIcon"
                    >
                      <path d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14">
                      </path>
                    </svg>
                    <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </header>
    </div>
    <div class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-n9dby4-MuiPaper-root">
      <div class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-1qbral5-MuiPaper-root"
           data-testid="Loading"
      >
        <h5 class="MuiTypography-root MuiTypography-h5 css-1oepjcl-MuiTypography-root">
          Loading, please wait...
        </h5>
        <span class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18li798-MuiCircularProgress-root"
              role="progressbar"
              style="width: 5rem; height: 5rem;"
        >
          <svg class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
               viewbox="22 22 44 44"
          >
            <circle class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    r="20.2"
                    fill="none"
                    stroke-width="3.6"
            >
            </circle>
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
